File: python/977_squares_of_a_sorted_array.py
def sortedSquares(nums: list[int]) -> list[int]:
    return sorted([num*num for num in nums])

def sortedSquaresFast(nums: list[int]) -> list[int]:
    n = len(nums)
    vec = [0] * n

    start = 0
    end = n - 1

    for idx in range(n - 1, -1, -1):
        if abs(nums[end]) > abs(nums[start]):
            vec[idx] = nums[end] * nums[end]
            end -= 1 
        else:
            vec[idx] = nums[start] * nums[start]
            start += 1
    return vec

File: javascript/977_squares_of_a_sorted_array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
function sortedSquares(nums) {
    return nums.map((num) => num * num).sort();
};

/**
 * @param {number[]} nums
 * @return {number[]}
 */
function sortedSquaresFast(nums) {
    const vec = new Array(nums.length);

    let start = 0;
    let end = nums.length - 1;
    
    for (let idx = nums.length - 1; idx > -1; idx--) {
        if (Math.abs(nums[end]) > Math.abs(nums[start])) {
            vec[idx] = nums[end] * nums[end];
            end--;
        } else {
            vec[idx] = nums[start] * nums[start];
            start++;
        }
    }
    return vec;
};

console.log(sortedSquaresFast([-4,-1,0,3,10])); // [0,1,9,16,100]
